README.md
# InSilicoVA
[![R-CMD-check](https://github.com/verbal-autopsy-software/InSilicoVA/workflows/R-CMD-check/badge.svg)](https://github.com/verbal-autopsy-software/InSilicoVA/actions) [![CRAN\_Status\_Badge](https://www.r-pkg.org/badges/version/InSilicoVA)](https://cran.r-project.org/package=InSilicoVA) [![](https://cranlogs.r-pkg.org/badges/InSilicoVA)](https://cran.r-project.org/package=InSilicoVA) [![](https://cranlogs.r-pkg.org/badges/grand-total/InSilicoVA?color=orange)](https://cran.r-project.org/package=InSilicoVA)

To install the package from CRAN, run the script: 
```r
install.packages("InSilicoVA")
```

To install the github testing version of InSilicoVA, run the script
```r
library(devtools)
devtools::install_github("verbal-autopsy-software/InSilicoVA", subdir = "InSilicoVA")
```

To use InSilicoVA, 'rJava' package is required and it requires Java 6 or above properly installed and linked to R. See <a href="Documents/Insilico-manual.pdf">the user manual</a> for more detail.

### Reference
Tyler H. McCormick, Zehang Richard Li, Clara Calvert, Amelia C. Crampin, Kathleen Kahn, and Samuel J. Clark. <a href="http://arxiv.org/abs/1411.3042">Probabilistic Cause-of-death Assignment using Verbal Autopsies</a>. _Journal of the American Statistical Association_ 111.515 (2016): 1036-1049.



### What's New
- [CRAN version](https://cran.r-project.org/web/packages/InSilicoVA/news/news.html) 
- [Current developer version](InSilicoVA/NEWS.md)
 

### Additional information when running large models
If you hit the error message compaining about java heap size, e.g. ``java.lang.OutOfMemoryError: Java heap space``

Try start a new R session and run this line of code **before** calling ``library(openVA)`` or ``library(InSilicoVA)``:

> options(java.parameters = c("-XX:+UseConcMarkSweepGC"))


InSilicoVA/NEWS.md
# InSilicoVA - changes
Version 1.4.0 (2022-09-25) 
==========================
+ Bug fix in `mapICD`.
+ New color palette that is colorblind friendly.

Version 1.3.5 (2022-01-10) 
==========================
+ `get.indiv` now can return posterior draws of P(cause | symptoms) for each individual.
+ New function `mapICD` to map ICD-10 codes to WHO 2016 cause list.

Version 1.3.4 (2021-08-01) 
==========================
+ Bug fix in missing data threshold for customized (non-WHO) input format.

Version 1.3.3 (2021-07-01) 
==========================
+ The default `subset2` option for removing impossible causes now include rules to avoid assigning pregnancy-related causes to deaths with i310o = Y (i.e., not pregnant).

Version 1.3.2 (2021-05-25) 
==========================
+ Minor fixes to pass Debian _R_CHECK_MATRIX_DATA.
 
Version 1.3.1 (2021-05-09) 
==========================
+ Minor bug fixes.
 
Version 1.3.0 (2020-06-03) 
==========================
+ New default option to specify impossible causes, "subset2", that deals with prematurity deaths.
+ Flexibility improvements in several functions.
 
Version 1.2.9 (2020-05-07) 
==========================
+ fix bug with subpop specification in WHO2016 input

Version 1.2.8 (2020-04-14)
==========================
+ Minor improvements in summary display.
+ Minor stability improvements in checking input formats.

Version 1.2.7 (2020-02-18)
==========================
+ Include system checks for Java version at least 7.0.

Version 1.2.6 (2020-02-17, on CRAN)
==========================
+ Include function to load the latest P(S|C) from InterVA-5 software

Version 1.2.6 (2018-12-10, on GitHub)
==========================
+ Bug fix for model with customized P(S|C)
+ Bug fix for scenario where all deaths due to external causes in WHO 2016 format.

Version 1.2.5 (2018-10-28)
==========================
+ Bug fix for subsetting data with (1) no external causes, and (2) only external causes.
+ Added checks for extra columns in physician debias function.

Version 1.2.4 (2018-09-12)
==========================
+ Add option to exclude impossible causes based on different rules.
+ The previous default rule to exclude impossible causes is now called 'subset'. It further include impossible causes screening based on the following symptoms: 
    + baby died within 24 hours; between 24 and 48 hours; between 48 hours and within one week; between one to four weeks. 
    + baby born dead.
    + woman 12-19 years old; 20-34 years old; 35-49 years old.
+ Bug fix symptoms not negated when absence is taken as the significant value. 
+ Bug fix that converts absence to missing for WHO 2016 input introduced in version 1.2.3.
+ Bug fix for individual probability calculation when no cause has positive probabilities.


Version 1.2.3 (2018-08-27)
==========================
+ Add codes to safely handle data containing NA values.
+ Fix the writing of warning and error log files for both WHO 2012 and 2016 formats, and allowed customized directory. 
 

Version 1.2.2 (2018-07-13)
========================== 
+ Fix typo in checking WHO 2016 input using InterVA5 rules, and update the check rules for external causes based on all symptoms associated with external causes (instead of using only the most deterministic one as in the previous versions).
+ Fix typo in shifting columns for WHO 2016 input.
+ Fix the issue that sometimes 0 acceptance rate is displayed when the chain is in fact moving.
+ Fix inconsistent names in InterVA5 default input.
+ Removed most of the cat() calls and replaced them with message().
+ Add subpop selection option for 'compare' plots.


Version 1.2.1 (2018-07-06)
==========================
+ Fix typo in processing WHO 2016 input.
+ Fix inconsistency from symptom name change (sk\_les and skin\_les).

Version 1.2.0 (2018-04-20)
========================== 
+ Add WHO 2016 support using InterVA5 input
+ Add option to allow group code in output
+ Add warning and error logs in the InSilico object, so that removed observations can be traced back
+ Make probbase consistent with InterVA-4.03 instead of 4.02. Changes should be minimum, but good to be consistent.
+ Allow user defined impossible causes to be removed


Version 1.1.5 (2017-11-17)
==========================
+ Fix issue for data without external causes.
+ Fix issue with print methods after last update.
+ Fix issue with neonate and child death assigned to impossible external causes.
+ Change data check steps to update symptoms to be missing instead of absence to introduce symptom dependence structures. 
+ Automatic remove impossible causes from CSMF based on subpopulation gender and age. 


Version 1.1.4 (2017-01-24)
========================== 
+ Fix issue with sub-population specification and output labels of non-standard InterVA-4 input.
 

Version 1.1.3 (2017-01-02)
==========================
+ Fix issue with label order in probbase output.
+ Allow input "Yes" to be denoted with either "y" or "Y".
+ Default data checking rules to be consistent with InterVA-4.03 instead of the previous version.
+ Fix typo in document.

    


InSilicoVA/R/mapICD.R
#' Map ICD-10 codes into the WHO 2016 cause list 
#' 
#' @param x a character object or a vector of ICD-10 codes  
#' @examples
#' mapICD("A90")
#' mapICD(c("A90", "C30"))
#' @export
#' 
mapICD <- function(x){

    icd10Map <- data.frame(
        rbind(
            c('Diarrhoeal diseases', 'A', 00, 09),
            c('Pulmonary tuberculosis', 'A', 15, 16),
            c('Other and unspecified infect dis', 'A', 17, 19),
            c('Other and unspecified infect dis', 'A', 20, 32),
            c('Tetanus (neonatorum)', 'A', 33, 33),
            c('Tetanus', 'A', 34, 35),
            c('Other and unspecified infect dis', 'A', 36, 36),
            c('Pertussis', 'A', 37, 37),
            c('Other and unspecified infect dis', 'A', 38, 38),
            c('Meningitis and encephalitis', 'A', 39, 39),
            c('Sepsis (non-obstetric)', 'A', 40, 41),
            c('Other and unspecified infect dis', 'A', 42, 89),
            c('Dengue fever', 'A', 90, 91),
            c('Haemorrhagic fever (non-dengue)', 'A', 92, 99),
            c('Other and unspecified infect dis', 'B', 00, 04),
            c('Measles', 'B', 05, 05),
            c('Other and unspecified infect dis', 'B', 06, 19),
            c('HIV/AIDS related death', 'B', 20, 24),
            c('Other and unspecified infect dis', 'B', 25, 49),
            c('Malaria', 'B', 50, 54),
            c('Other and unspecified infect dis', 'B', 55, 99),
            c('Oral neoplasms', 'C', 00, 06),
            c('Other and unspecified neoplasms', 'C', 07, 14),
            c('Digestive neoplasms', 'C', 15, 26),
            c('Respiratory neoplasms', 'C', 30, 39),
            c('Other and unspecified neoplasms', 'C', 40, 49),
            c('Breast neoplasms', 'C', 50, 50),
            c('Reproductive neoplasms (female)', 'C', 51, 58),
            c('Reproductive neoplasms (male)', 'C', 60, 63),
            c('Other and unspecified neoplasms', 'C', 60, 96),
            c('Other and unspecified neoplasms', 'D', 00, 48),
            c('Severe anaemia', 'D', 50, 56),
            c('Sickle cell with crisis', 'D', 57, 57),
            c('Severe anaemia', 'D', 58, 64),
            c('Other and unspecified NCD', 'D', 65, 89),
            c('Other and unspecified NCD', 'E', 00, 07),
            c('Diabetes mellitus', 'E', 10, 14),
            c('Other and unspecified NCD', 'E', 15, 35),
            c('Severe malnutrition', 'E', 40, 46),
            c('Other and unspecified NCD', 'E', 50, 90),
            c('Other and unspecified NCD', 'F', 00, 99),
            c('Meningitis and encephalitis', 'G', 00, 05),
            c('Other and unspecified NCD', 'G', 06, 37),
            c('Epilepsy', 'G', 40, 41),
            c('Other and unspecified NCD', 'G', 50, 99),
            c('Other and unspecified NCD', 'H', 00, 95),
            c('Other and unspecified cardiac dis', 'I', 00, 15),
            c('Acute cardiac disease', 'I', 20, 25),
            c('Other and unspecified cardiac dis', 'I', 26, 52),
            c('Stroke', 'I', 60, 69),
            c('Other and unspecified cardiac dis', 'I', 70, 99),
            c('Acute resp infect incl pneumonia', 'J', 00, 22),
            c('Other and unspecified NCD', 'J', 30, 39),
            c('Chronic obstructive pulmonary dis', 'J', 40, 44),
            c('Asthma', 'J', 45, 46),
            c('Other and unspecified NCD', 'J', 47, 99),
            c('Other and unspecified NCD', 'K', 00, 31),
            c('Other and unspecified NCD', 'K', 35, 38),
            c('Other and unspecified NCD', 'K', 40, 69),
            c('Liver cirrhosis', 'K', 70, 76),
            c('Other and unspecified NCD', 'K', 77, 93),
            c('Other and unspecified NCD', 'L', 00, 99),
            c('Other and unspecified NCD', 'M', 00, 99),
            c('Other and unspecified NCD', 'N', 00, 16),
            c('Renal failure', 'N', 17, 19),
            c('Other and unspecified NCD', 'N', 20, 99),
            c('Ectopic pregnancy', 'O', 00, 00),
            c('Other and unspecified maternal CoD', 'O', 01, 02),
            c('Abortion-related death', 'O', 03, 08),
            c('Pregnancy-induced hypertension', 'O', 10, 16),
            c('Other and unspecified maternal CoD', 'O', 20, 45),
            c('Obstetric haemorrhage', 'O', 46, 46),
            c('Other and unspecified maternal CoD', 'O', 47, 62),
            c('Obstructed labour', 'O', 63, 66),
            c('Obstetric haemorrhage', 'O', 67, 67),
            c('Other and unspecified maternal CoD', 'O', 68, 70),
            c('Ruptured uterus', 'O', 71, 71),
            c('Obstetric haemorrhage', 'O', 72, 72),
            c('Other and unspecified maternal CoD', 'O', 73, 75.2),
            c('Pregnancy-related sepsis', 'O', 75.3, 75.3),
            c('Other and unspecified maternal CoD', 'O', 75.4, 84),
            c('Pregnancy-related sepsis', 'O', 85, 85),
            c('Other and unspecified maternal CoD', 'O', 86, 98.9),
            c('Anaemia of pregnancy', 'O', 99.0, 99.0),
            c('Other and unspecified maternal CoD', 'O', 99.1, 99.8),
            c('Other and unspecified neonatal CoD', 'P', 00, 04),
            c('Prematurity', 'P', 05, 07),
            c('Other and unspecified neonatal CoD', 'P', 08, 15),
            c('Birth asphyxia', 'P', 20, 22),
            c('Neonatal pneumonia', 'P', 23, 25),
            c('Other and unspecified neonatal CoD', 'P', 26, 35),
            c('Neonatal sepsis', 'P', 36, 36),
            c('Other and unspecified neonatal CoD', 'P', 37, 94),
            c('Fresh stillbirth', 'P', 95, 95),
            c('Macerated stillbirth', 'P', 95, 95),  ## problem (same as above) CoD will always be Macerated (never Fresh)
            c('Other and unspecified neonatal CoD', 'P', 96, 96),
            c('Congenital malformation', 'Q', 00, 99),
            c('Other and unspecified NCD', 'R', 00, 09),
            c('Acute abdomen', 'R', 10, 10),
            c('Other and unspecified NCD', 'R', 11, 94),
            c('Unknown', 'R', 95, 99),
            c('Other and unspecified external CoD', 'S', 00, 99),
            c('Other and unspecified external CoD', 'T', 00, 99),
            c('Road traffic accident', 'V', 01, 89),
            c('Other transport accident', 'V', 90, 99),
            c('Accid fall', 'W', 00, 19),
            c('Other and unspecified external CoD', 'W', 20, 64),
            c('Accid drowning and submersion', 'W', 65, 74),
            c('Other and unspecified external CoD', 'W', 75, 99),
            c('Accid expos to smoke fire & flame', 'X', 00, 19),
            c('Contact with venomous plant/animal', 'X', 20, 29),
            c('Exposure to force of nature', 'X', 30, 39),
            c('Accid poisoning & noxious subs', 'X', 40, 49),
            c('Other and unspecified external CoD', 'X', 50, 59),
            c('Intentional self-harm', 'X', 60, 84),
            c('Assault', 'X', 85, 99),
            c('Assault', 'Y', 00, 09),
            c('Other and unspecified external CoD', 'Y', 10, 98)
        ), stringsAsFactors = FALSE
    )
    
    out <- rep(NA, length(x))
    for(jj in 1:length(x)){
        icd10Code <- x[jj]
        parsedCode <- cbind(substr(icd10Code, 1, 1),
                        substr(icd10Code, 2, 3),
                        substr(icd10Code, 5, 5))
        tmp <- 'Other (not in InterVA-5 cause list)'
        for (i in 1:nrow(icd10Map)) {  

            if (parsedCode[1] != icd10Map[i, 2]) next

            start <- as.numeric(icd10Map[i, 3])

            stop <- as.numeric(icd10Map[i, 4])
            if (round(stop) == stop) {
                stop <- as.numeric(paste0(stop, '.9'))
                }
            
            numCode <- paste0(parsedCode[2], '.', parsedCode[3])
            numCode <- as.numeric(numCode)
            if (start <= numCode & numCode <= stop) {
                if(icd10Map[i, 1] == "Tetanus (neonatorum)") icd10Map[i, 1] = "Tetanus"
                if(icd10Map[i, 1] == "Reproductive neoplasms (female)") icd10Map[i, 1] = "Reproductive neoplasms"
                if(icd10Map[i, 1] == "Reproductive neoplasms (male)") icd10Map[i, 1] = "Reproductive neoplasms"
                tmp <- icd10Map[i, 1]
                break
            }
        }
        out[jj] <- tmp
    }
    
    return(out)
}